File: motor/src/grootboek.rs
/* Het grootboek: dubbel boekhouden op de cent, exact zoals server/kern/pay.
   Bedragen zijn i64 CENTEN — geen float, dus geen afrondingsdrift. Elke
   beweging is VAN een rekening NAAR een rekening; geld ontstaat nooit uit het
   niets (opladen komt van extern:oplaad, uitbetalen gaat naar
   extern:uitbetaald). De som van alle saldi is altijd exact nul, en niemand
   behalve de extern-rekeningen mag rood staan. */
use crate::json::Json;
use crate::rng;
use std::collections::HashMap;
use std::collections::VecDeque;

pub const MIN_CENTEN: i64 = 1;
pub const MAX_CENTEN: i64 = 500_000;
pub const WEERGAVE_CAP: usize = 50_000;

#[derive(Clone)]
pub struct Boeking {
    pub id: String,
    pub van: String,
    pub naar: String,
    pub centen: i64,
    pub soort: String,
    pub oms: String,
    pub ref_: Option<String>,
    pub at: u64,
}

impl Boeking {
    pub fn to_json(&self) -> Json {
        let mut o = Json::obj();
        o.set("id", Json::Str(self.id.clone()))
            .set("van", Json::Str(self.van.clone()))
            .set("naar", Json::Str(self.naar.clone()))
            .set("centen", Json::Num(self.centen as f64))
            .set("soort", Json::Str(self.soort.clone()))
            .set("oms", Json::Str(self.oms.clone()))
            .set("ref", self.ref_.clone().map(Json::Str).unwrap_or(Json::Null))
            .set("at", Json::Num(self.at as f64));
        o
    }
}

pub struct BoekArgs<'a> {
    pub van: &'a str,
    pub naar: &'a str,
    pub centen: i64,
    pub soort: &'a str,
    pub oms: &'a str,
    pub ref_: Option<String>,
}

pub struct Ledger {
    pub saldi: HashMap<String, i64>,
    pub boekingen: VecDeque<Boeking>, // front = nieuwste (unshift-semantiek)
}

impl Ledger {
    pub fn new() -> Ledger {
        Ledger { saldi: HashMap::new(), boekingen: VecDeque::new() }
    }

    pub fn saldo_van(&self, rek: &str) -> i64 {
        *self.saldi.get(rek).unwrap_or(&0)
    }

    /* De kern: een boeking. Weigert onzin (bedrag buiten bereik, van==naar) en
       laat niemand behalve de extern-rekeningen onder nul zakken. */
    pub fn boek(&mut self, a: BoekArgs) -> Result<Boeking, (u16, String)> {
        let c = a.centen;
        if c < MIN_CENTEN || c > MAX_CENTEN {
            return Err((400, "Dat bedrag kan niet.".into()));
        }
        if a.van.is_empty() || a.naar.is_empty() || a.van == a.naar {
            return Err((400, "Van en naar kloppen niet.".into()));
        }
        if !a.van.starts_with("extern:") && self.saldo_van(a.van) < c {
            return Err((402, "Onvoldoende saldo.".into()));
        }
        *self.saldi.entry(a.van.to_string()).or_insert(0) -= c;
        *self.saldi.entry(a.naar.to_string()).or_insert(0) += c;
        let rij = Boeking {
            id: rng::id("PB"),
            van: a.van.to_string(),
            naar: a.naar.to_string(),
            centen: c,
            soort: a.soort.to_string(),
            oms: schoon(a.oms, 120),
            ref_: a.ref_,
            at: rng::nu_ms(),
        };
        self.boekingen.push_front(rij.clone());
        if self.boekingen.len() > WEERGAVE_CAP {
            self.boekingen.pop_back();
        }
        Ok(rij)
    }

    /* Rauw toepassen (schaduw-modus): herspeelt een boeking van de autoritaire
       JS-engine ZONDER saldo-guard — het herspeelt een al-genomen beslissing,
       het neemt hem niet opnieuw. Zo blijft de mirror in lockstep, ook als
       spiegelingen door de storm net anders geordend binnenkomen. */
    pub fn apply_raw(&mut self, a: BoekArgs) -> Boeking {
        *self.saldi.entry(a.van.to_string()).or_insert(0) -= a.centen;
        *self.saldi.entry(a.naar.to_string()).or_insert(0) += a.centen;
        let rij = Boeking {
            id: rng::id("PB"),
            van: a.van.to_string(),
            naar: a.naar.to_string(),
            centen: a.centen,
            soort: a.soort.to_string(),
            oms: schoon(a.oms, 120),
            ref_: a.ref_,
            at: rng::nu_ms(),
        };
        self.boekingen.push_front(rij.clone());
        if self.boekingen.len() > WEERGAVE_CAP {
            self.boekingen.pop_back();
        }
        rij
    }

    /* De sluitcontrole: som van alle saldi is nul, en niemand staat rood. Dit
       is de waarheid waar /api/pay/gezond op afgaat. */
    pub fn sluitcontrole(&self) -> (bool, i64, Vec<String>) {
        let mut som: i64 = 0;
        let mut rood = Vec::new();
        for (rek, &c) in &self.saldi {
            som += c;
            if !rek.starts_with("extern:") && c < 0 {
                rood.push(rek.clone());
            }
        }
        (som == 0 && rood.is_empty(), som, rood)
    }
}

/* Kort schoonmaken van vrije tekst (zoals `schoon` in de Node-kern): knip af,
   gooi controltekens weg. */
pub fn schoon(s: &str, max: usize) -> String {
    s.chars()
        .filter(|c| !c.is_control())
        .take(max)
        .collect::<String>()
        .trim()
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dubbel_boekhouden_blijft_op_nul() {
        let mut g = Ledger::new();
        // opladen: extern:oplaad -> lid:NEVEL
        g.boek(BoekArgs { van: "extern:oplaad", naar: "lid:NEVEL", centen: 50000, soort: "oplaad", oms: "", ref_: None }).unwrap();
        // sturen: lid:NEVEL -> lid:MIST
        g.boek(BoekArgs { van: "lid:NEVEL", naar: "lid:MIST", centen: 25000, soort: "p2p", oms: "", ref_: None }).unwrap();
        let (klopt, som, rood) = g.sluitcontrole();
        assert!(klopt, "grootboek moet sluiten");
        assert_eq!(som, 0);
        assert!(rood.is_empty());
        assert_eq!(g.saldo_van("lid:NEVEL"), 25000);
        assert_eq!(g.saldo_van("lid:MIST"), 25000);
        assert_eq!(g.saldo_van("extern:oplaad"), -50000);
    }

    #[test]
    fn lid_kan_niet_onder_nul() {
        let mut g = Ledger::new();
        let r = g.boek(BoekArgs { van: "lid:LEEG", naar: "lid:X", centen: 100, soort: "p2p", oms: "", ref_: None });
        assert!(matches!(r, Err((402, _))));
        // saldi ongemoeid, sluit nog
        assert!(g.sluitcontrole().0);
    }

    #[test]
    fn onzin_bedrag_geweigerd() {
        let mut g = Ledger::new();
        assert!(matches!(g.boek(BoekArgs { van: "extern:oplaad", naar: "lid:X", centen: 0, soort: "x", oms: "", ref_: None }), Err((400, _))));
        assert!(matches!(g.boek(BoekArgs { van: "extern:oplaad", naar: "lid:X", centen: 600_000, soort: "x", oms: "", ref_: None }), Err((400, _))));
        assert!(matches!(g.boek(BoekArgs { van: "lid:X", naar: "lid:X", centen: 100, soort: "x", oms: "", ref_: None }), Err((400, _))));
    }
}

File: server/kern/bank/grootboek.js
/* RTG Bank, deel "grootboek": de boekhoudmotor zelf + de cutover-naad naar de
   Rust-motor. Afgesplitst uit kern/bank/index.js zodat de orkestrator klein blijft.

   De regels van dit grootboek (dezelfde tucht als RTG Pay):
   - elke beweging is een dubbele boeking VAN -> NAAR; de som van alle saldi is
     altijd exact nul; niemand zakt onder zijn bodem (rood-staan-limiet per soort).

   Cutover (RTG_MOTOR_GELD=motor): de Rust-motor wordt de autoritatieve saldi-store.
   Anders dan pay doet de motor een RAUWE apply -- de rijke bank-guard (rekening
   bestaat / bevroren / rood-staan-bodem) leunt op de rekening-metadata die in JS
   woont, dus die guard blijft hier. Daarom: guard in JS, dan rauw naar de motor,
   dan spiegelen. Een serialisatie-slot om alle bank-schrijfacties heen sluit de
   TOCTOU-race waarin twee gelijktijdige boekingen dezelfde verouderde bodem-check
   passeren. Standaard uit (geldModus 'schaduw') -> puur de synchrone JS-guard. */
'use strict';

const { vingerafdruk } = require('../pay/vingerafdruk');

module.exports = (g) => {
  const { MIN_CENTEN, MAX_CENTEN, saldi, grootboek, saldoVan, rekMeta, isExtern,
    bodem, id, schoon, nu, save, d, geldModus, motorklant, bordSeintje } = g;
  const betalingenUit = process.env.RTG_BETALEN_UIT === '1';
  const uitFout = () => ({ status: 503,
    error: 'Betalen staat bewust uitgeschakeld. Er is niets afgeschreven.', code: 'betalingen-uit' });

  /* De grootboek-guard: de VOLLEDIGE regelcontrole (bedrag, bestaan, bevroren,
     bodem/rood-staan). Retourneert een fout-object of null (mag door). Dit is de
     enige plek waar de bank-policy leeft -- in schaduw-modus draait hij synchroon
     vóór het toepassen; in motor-modus draait hij in JS vóór de rauwe apply op de
     motor (die de metadata niet kent). */
  function guardCheck({ van, naar, c }) {
    if (!Number.isFinite(c) || c < MIN_CENTEN || c > MAX_CENTEN) return { status: 400, error: 'Dat bedrag kan niet.' };
    if (!van || !naar || van === naar) return { status: 400, error: 'Van en naar kloppen niet.' };
    if (!isExtern(van)) { const mv = rekMeta(van); if (!mv) return { status: 404, error: 'De rekening bestaat niet.' }; if (mv.bevroren) return { status: 423, error: 'Deze rekening is bevroren.' }; }
    if (!isExtern(naar) && !rekMeta(naar)) return { status: 404, error: 'De tegenrekening bestaat niet.' };
    if (saldoVan(van) - c < bodem(van)) return { status: 402, error: 'Onvoldoende saldo of rood-staan-ruimte.' };
    return null;
  }
  /* Een AL-goedgekeurde boeking toepassen op de saldi + het grootboek (geen guard
     meer). Gedeeld door de sync-guard (boek) en de motor-spiegel (boekMotor). */
  function pasToe(rij) {
    saldi()[rij.van] = saldoVan(rij.van) - rij.centen;
    saldi()[rij.naar] = saldoVan(rij.naar) + rij.centen;
    grootboek().unshift(rij);
    if (grootboek().length > 100000) grootboek().pop();  // weergavecap; de saldi zijn de waarheid
    save();
    bordSeintje();
  }

  /* De synchrone grootboekmotor. Boekt van -> naar, bewaakt de bodem en de dubbele
     boeking. In motor-modus mag dit NIET: dan is de motor het autoritatieve saldi-
     grootboek en moet alles via boekAsync (fail-closed, luid -- nooit stil een
     tweede grootboek naast de motor). */
  function boek({ van, naar, centen, soort, oms, ref }) {
    if (betalingenUit) return uitFout();
    if (geldModus === 'motor') {
      throw new Error('bank.boek (synchroon) is niet toegestaan in RTG_MOTOR_GELD=motor; gebruik boekAsync.');
    }
    const c = Math.round(Number(centen));
    const fout = guardCheck({ van, naar, c });
    if (fout) return fout;
    const rij = { id: id('BB'), van, naar, centen: c, soort: soort || 'boeking', oms: schoon(oms, 140), ref: ref || null, at: nu() };
    pasToe(rij);
    return { ok: true, boeking: rij };
  }

  /* De serialisatie-slot voor de bank-schrijfacties in motor-modus. De guard leest
     de spiegel-saldi en pas ná de motor-bevestiging past hij ze aan; zonder slot
     konden twee gelijktijdige overboekingen van dezelfde rekening allebei dezelfde
     verouderde bodem-check passeren (TOCTOU) en samen door de bodem zakken. Door
     alle bank-schrijfacties door één belofte-keten te trekken ziet elke guard altijd
     de laatst toegepaste stand. */
  let schrijfKeten = Promise.resolve();
  function metSlot(werk) {
    const uit = schrijfKeten.then(werk, werk);
    schrijfKeten = uit.then(() => {}, () => {}); // de keten breekt nooit op een fout
    return uit;
  }
  /* De async boeking: HET choke-point voor de cutover. In schaduw-modus is dit exact
     de sync-guard (gewoon awaitbaar) -- geen gedragsverandering. In motor-modus draait
     de JS-guard (metadata!) eerst; pas als die doorlaat gaat de boeking rauw naar de
     motor (autoriteit voor de saldi). Bevestigt de motor, dan spiegelt JS dezelfde
     regel; weigert of hapert de motor, dan verandert er NIETS aan de spiegel. Alle
     motor-schrijfacties lopen door het slot (geen TOCTOU op de bodem). */
  async function boekAsync(args) {
    if (betalingenUit) return uitFout();
    if (geldModus !== 'motor') return boek(args);
    return metSlot(() => boekMotor(args));
  }
  async function boekMotor({ van, naar, centen, soort, oms, ref, economischeSleutel }) {
    const c = Math.round(Number(centen));
    const fout = guardCheck({ van, naar, c });
    if (fout) return fout;
    const r = await motorklant.bankBoek({ van, naar, centen: c, soort, oms, ref, economischeSleutel });
    if (!r || r.error) return { status: (r && r.status) || 502, error: (r && r.error) || 'Motor onbereikbaar.' };
    const b = r.boeking;
    const rij = { id: b.id, van: b.van, naar: b.naar, centen: Math.round(Number(b.centen)), soort: b.soort || 'boeking', oms: b.oms || '', ref: b.ref || null, at: b.at || nu() };
    if (economischeSleutel) {
      const sv = Math.round(Number(r.saldoVan)), sn = Math.round(Number(r.saldoNaar));
      if (!Number.isFinite(sv) || !Number.isFinite(sn))
        return { status: 502, error: 'Motor bevestigde geen actuele banksaldi; de spiegel blijft ongemoeid.' };
      saldi()[rij.van] = sv; saldi()[rij.naar] = sn;
      if (!grootboek().some(x => x && x.id === rij.id)) {
        grootboek().unshift(rij);
        if (grootboek().length > 100000) grootboek().pop();
      }
      save(); bordSeintje();
      return { ok: true, boeking: rij, herhaald: !!r.herhaald };
    }
    pasToe(rij);
    return { ok: true, boeking: rij };
  }

  /* Herstart-reconcile (cutover): bij het opstarten in motor-modus is de motor de
     autoriteit voor de bank-saldi, dus de JS-spiegel neemt zijn saldi over uit de
     motor-snapshot i.p.v. uit zijn eigen (mogelijk verouderde) snapshot. Zo start de
     spiegel altijd in lockstep, ook na een crash. Vereist RTG_MOTOR_SALDI=1 op de
     motor. No-op buiten motor-modus. */
  async function reconcileVanMotor() {
    if (geldModus !== 'motor') return { ok: true, overgeslagen: true };
    const r = await motorklant.bankSaldiSnapshot();
    if (!r || r.error) return { ok: false, error: (r && r.error) || 'Geen saldi van de motor.' };
    const nieuw = {};
    for (const k in r.saldi) {
      if (!Object.prototype.hasOwnProperty.call(r.saldi, k)) continue;
      const v = Math.round(Number(r.saldi[k]) || 0);
      if (v !== 0) nieuw[k] = v; // nul-saldi laten we weg (schone spiegel)
    }
    d().bankSaldi = nieuw;
    save();
    let som = 0; for (const k in nieuw) som += nieuw[k];
    return { ok: true, rekeningen: Object.keys(nieuw).length, som };
  }

  // de sluitcontrole: som van alle saldi is nul, en niemand zit onder zijn bodem
  function sluitcontrole() {
    let som = 0; const onderBodem = [];
    for (const [rek, c] of Object.entries(saldi())) { som += c; if (c < bodem(rek)) onderBodem.push(rek); }
    return { klopt: som === 0 && !onderBodem.length, som, onderBodem };
  }

  /* Drift-stand voor het statusbord (cutover): vergelijkt de JS-spiegel met de
     Rust-motor -- niet alleen de som maar ook een vingerafdruk over ALLE bank-saldi,
     zodat per-rekening-drift die de som mist er alsnog uitkomt. Alleen op een
     statusbord-poll (nooit in het warme geld-pad); byte-voor-byte gelijk aan de motor. */
  const motorStand = {
    aan: geldModus === 'motor', modus: geldModus,
    stand: async () => {
      const s = sluitcontrole();
      const jsAfdruk = vingerafdruk(saldi());
      if (geldModus !== 'motor') return { modus: geldModus, jsSom: s.som, jsVingerafdruk: jsAfdruk };
      const r = await motorklant.bankSaldiSnapshot();
      if (!r || r.error) return { modus: geldModus, jsSom: s.som, jsVingerafdruk: jsAfdruk, fout: (r && r.error) || 'geen motor-saldi' };
      let motorSom = 0; for (const k in r.saldi) motorSom += Math.round(Number(r.saldi[k]) || 0);
      const motorAfdruk = vingerafdruk(r.saldi);
      return { modus: geldModus, jsSom: s.som, motorSom, gelijk: motorSom === s.som,
        jsVingerafdruk: jsAfdruk, motorVingerafdruk: motorAfdruk, gelijkAlle: jsAfdruk === motorAfdruk };
    },
  };

  return { guardCheck, pasToe, boek, boekAsync, reconcileVanMotor, sluitcontrole, motorStand };
};
